service/history/workflow/update/registry.go

```diff
@@ -136,49 +139,85 @@ func (r *RegistryImpl) CreateOutgoingMessages(startedEventID int64) ([]*protocol
 	return updMessages, nil
 }
 
+func (r *RegistryImpl) ValidateIncomingMessages(messages []*protocolpb.Message) error {
+	r.Lock()
+	defer r.Unlock()
+
+	// make a copy of updates for validation so that it does not alter the state which will be used
+	// by a retry of failed workflow task
+
+	updates := make(map[string]*Update, len(r.updates))
+	for k, v := range r.updates {
+		updates[k] = &Update{
+			state:              v.state,
+			request:            v.request,
+			messageID:          v.messageID,
+			protocolInstanceID: v.protocolInstanceID,
+			out:                nil, // we don't need it for validation
+		}
+	}
+
+	_, err := processIncomingMessages(updates, messages)
+	return err
+}
+
 func (r *RegistryImpl) ProcessIncomingMessages(messages []*protocolpb.Message) error {
 	r.Lock()
 	defer r.Unlock()
+
+	closedUpdates, err := processIncomingMessages(r.updates, messages)
+	if err != nil {
+		return err
+	}
+
+	// notify result to caller
+	for _, upd := range closedUpdates {
+		upd.notifyResult()
+	}
+	return nil
+}
+
+func processIncomingMessages(updates map[string]*Update, messages []*protocolpb.Message) ([]*Update, error) {
+	var closedUpdates []*Update
```

Comment:
I don't like this name. `completedUpdates` seems better, even if it contains both successfully completed and rejected updates.

Reply:
we already have closed workflow / closed activity, I think it is more clear to convey the meaning.

Reply:
Yeah, and I don't like it there neither. But ok, that probably just me.